README.md
# WatchLyst

<p align="center">
  <img src="https://i.imgur.com/WUXKZzZ.png" alt="WatchLyst Logo"/>
  <p align="center">Thanks <a href="https://twitter.com/MarcelSteak3">@MarcelSteak3</a> for the icon!</p>
  <hr />
</p>

A Discord bot which allows server staff to create a list of troublesome users, and will alert when one of the users join the server.

![GIF Example](https://i.imgur.com/q1mi4X8.gif "GIF Example")

### [Invite WatchLyst to your server](https://discord.com/api/oauth2/authorize?client_id=765240772781932555&permissions=2052&scope=bot%20applications.commands)

# Usage

### Commands

- `/add [User ID*] [Reason]` Add a user to the server's WatchLyst

- `/remove [User ID*]` Remove a user from the server's WatchLyst

- `/update [User ID*] [Reason]` Update a user's information in the server's WatchLyst

- `/list [Page]` Display a list of users in the server's WatchLyst

- `/check [User ID*]` Show information about a user in the server's WatchLyst

- `/help` Display a list of WatchLyst commands (Also usable in DMs)

- `/config [Channel ID] [Role ID] [Toggle Pings]` (Admin only) Configurate a channel that WatchLyst will send alerts to, a role that will have access to the WatchLyst commands, and whether or not to ping the role when a user joins the server. Leave all arguments blank to display or reset the configuration.

### Events

- **When WatchLyst is added to a new server** -> WatchLyst will send the owner a message and create new configuration for the server.

- **When WatchLyst is removed from a server** -> WatchLyst will remove the server's configuration.

- **When a listed user joins or leaves** -> WatchLyst will send an alert.

- **When a listed user is banned** -> WatchLyst will send an alert and unlist the user.

- **When an assigned channel is deleted** -> The channel will be removed from the server's WatchLyst configuration.

- **When an assigned role is deleted** -> The role will be removed from the server's WatchLyst configuration.

CHANGELOG.md
### 2.0.4 - April 17 2023
- Updated packages
- Bot status now resets once an hour (Fixes a bug where the status would disappear after starting bot)

### 2.0.3 - March 15 2023
- Updated icon credit in README.md
- Small change in `/config` description
- Now calls avatarURL instead of hardcoded URL
- Updated packages and discord.js to 14.8.0
- Properly added ESLint and linted all files

### 2.0.2 - February 22 2023
- Removed unnecessary intents
- Small change in `/config` description

### 2.0.1 - February 13 2023
- Fixed a bug where the bot would crash when it leaves a server
- Lists now display the newest entries first

### 2.0.0 - February 12 2023
- Completely rewrote the bot using TypeScript and [discordx](https://discordx.js.org/)
- Updated to discord.js v14
- Now using the [Sequelize](https://sequelize.org/) ORM
- Now using environmental variables to store credentials
- Replaced all regular commands with Slash commands
- WatchLyst now requires less permissions to function (Send Messages and Ban Members)
- Listed users will be displayed with their username and ID instead of just ID
- WatchLyst now handles Snowflake IDs through the Discord API instead of as strings. Non-existent IDs will be rejected
- Changed `/setup` to `/config [Channel] [Role] [Toggle Ping]` and removed `/channel` `/role` `/toggleping`
- Made it possible to remove assigned channel and role by using `/config` without arguments
- Added navigation buttons to `/list` to flip between pages
- Increased the maximum amount of characters for a listed user's reason from 512 to 999
- Added `/update [User ID*] [New description]` as a way to update existing users
- A message will be sent when a listed user leaves the server. That will not remove the listed user from the server's WatchLyst
- When a listed user is banned from the server, they will be automatically unlisted
- When an assigned role or channel are deleted in a server, WatchLyst will automatically remove them from the server's configuration and notify the server's owner
- A server's WatchLyst information will automatically be destroyed when WatchLyst is removed from the server (but not its listed users)
- Error messages will only be displayed to the user who sent the command
- Fixed an issue where IDs not 18 characters in length would not be registered

### 1.2.0 - March 11 2022
- New logo (Thanks [@MarcelSteak2](https://twitter.com/MarcelSteak2)!)
- Upgraded to discord.js v13
- Code rewrite / cleanup
- Added `!w check`, which displays details about a specific user in the server's WatchLyst
- Added `!w clearbans`, which removes all of the banned users from the server's WatchLyst
- Add command now displays the provided reason
- Changed some colors around to match the new logo
- Fixed a bug where the aliases for `!w toggleping` was not functioning
- Fixed a bug where command parameters would not work if seperated by newlines or extra spaces
- Add command no longer replaces an apostophe (') with a tilted apostophe (’)
- Improved error messages

### 1.1.1 - January 1 2021
- Fixed a bug where WatchLyst would check for a bot-breaking character only once instead of all instances in an entry's reason
- When a user encounters an issue, WatchLyst now directs them to the GitHub

### 1.1.0 - December 24 2020
- Added pages to the list command. Can be used by adding a number after typing the command (May be replaced in the future with reactions as a way to flip between the pages)
- Added a check to see if a user is already in the server when adding that user
- Limited listing reason from 1,000 characters to 512 for future-proofing

### 1.0.3 - November 24 2020
- Reenabled banned user check (New method)
- The 'OK' emoji is now in a lighter shade of green

### 1.0.2 - November 21 2020
- Added an option to toggle on/off pinging the assigned role when a listed user joins the server with `!w toggleping` (Default: Off)
- Added emojis to most of the embed messages
- The 'You don't have permission to use this' and 'Only an admin may use this command' errors will now be deleted after 5 seconds

### 1.0.1 - November 20 2020
- Disabled banned users check when listing a new user due to reports of false positives
- Updated description for when a bot sends a message to the server's owner for the first time

### 1.0.0 - November 19 2020
- Initial release

src/commands/add.ts
import { ColorResolvable, CommandInteraction, EmbedBuilder, GuildMemberRoleManager, PermissionsBitField, User, ApplicationCommandOptionType } from "discord.js";
import { Discord, Slash, SlashOption } from "discordx";
import { watchlystConfig, emojiList } from "../config/botConfig.js";
import Servers from "../database/models/Servers.js";
import UserList from "../database/models/UserList.js";

@Discord()
export class Add {
  @Slash({ description: "Add a user to the server's WatchLyst", name: "add", dmPermission: false })
  async add(
    @SlashOption({
      type: ApplicationCommandOptionType.User,
      name: "user_id",
      description: "The ID of the user being added (17-19 digits long)",
      required: true,
    })
      user: User,
    @SlashOption({
      type: ApplicationCommandOptionType.String,
      name: "reason",
      description: "The reason for adding the user",
      required: false,
      maxLength: 999,
    })
      reason: string,
      interaction: CommandInteraction): Promise<void> {
    const serverQuery = await Servers.findOne({ where: { server_id: interaction.guild?.id } });
    const isAdmin = (interaction.member?.permissions as PermissionsBitField).has(PermissionsBitField.Flags.Administrator);
    const hasRole = (interaction.member?.roles as GuildMemberRoleManager).cache.has(serverQuery?.role_id as string);
    if (!isAdmin && !hasRole) {
      return void await interaction.reply({ ephemeral: true, embeds: [errorEmbed("You don't have permission to use this.")] });
    }

    try {
      const userTag = (await interaction.guild?.bans.fetch(user.id))?.user.tag;
      return void await interaction.reply({ ephemeral: true, embeds: [errorEmbed(`${userTag} (${user.id}) is already banned in the server.`)] });
    } catch { /* empty */ }
    try {
      const userTag = (await interaction.guild?.members.fetch(user.id))?.user.tag;
      return void await interaction.reply({ ephemeral: true, embeds: [errorEmbed(`${userTag} (${user.id}) is currently in the server.`)] });
    } catch { /* empty */ }

    const userQuery = await UserList.findOne({ where: { server_id: interaction.guild?.id, user_id: user.id } });
    if (userQuery !== null) {
      const alreadyInServerEmbed = errorEmbed(`${user.tag} (${user.id}) is already listed in the server's WatchLyst.`);
      alreadyInServerEmbed.addFields({ name: "Reason for adding:", value: userQuery.reason ?? "No reason provided." });
      alreadyInServerEmbed.setFooter({ text: "Note: You can use `/update [User ID*] [Reason]` to update the user." });
      return void await interaction.reply({ ephemeral: true, embeds: [alreadyInServerEmbed] });
    }

    await UserList.create({ user_id: user.id, server_id: interaction.guild?.id, reason: reason ?? null, added_by: interaction.user.tag })
      .then(() => interaction.reply({ embeds: [userAddedEmbed(user, reason)] }));
  }
}

const userAddedEmbed = (user: User, reason: string | undefined) => new EmbedBuilder()
  .setColor(watchlystConfig.colorGreen as ColorResolvable)
  .setDescription(`${emojiList.ok} ${user.tag} (${user.id}) has been added to the server's WatchLyst.`)
  .setFields(
    { name: "Reason for adding:", value: reason ?? "No reason provided." },
  );

const errorEmbed = (description: string) => new EmbedBuilder()
  .setColor(watchlystConfig.colorRed as ColorResolvable)
  .setDescription(`${emojiList.error} Error: ${description}`);
